### PR TITLE
Add document for diagnosis's design.

### PR DESCRIPTION
**Pull request by BalaBalaYi**
### What changes were proposed in this pull request?

n/a

### Why are the changes needed?

Add design document.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

No.

### CODE REVIEW

**Review by majieyue — APPROVED** [review]
lgtm

File: docs/design/proactive-diagnosis.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+### Diagnosis Agent
+The diagnosis-agent is implemented and runs on the worker node, with the main 
+responsibilities of driving data collection on the training worker side, 
+executing some basic diagnostic logic, and reporting necessary information to 
+Master.
+
+<img src="../figures/diagnosis_agent_arc.png" alt="Diagnosis Agent">
+
+### Other Concepts
+#### Metric Collection
```
Comment (workingloong) [inline docs/design/proactive-diagnosis.md:68]: Do you consider using Flight Recorder of torch >= 2.4.0 to diagnose hang?

File: docs/design/proactive-diagnosis.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+This part of the work will have a long cycle and work in progress for now...
+
+### Training Hang Detection(Basic)
+
+```
+prerequisite: XPU-Timer is deployed and in used.
+```
+
+Currently, the basic capability is to determine whether a Hang situation exists.
```
Comment (workingloong) [inline docs/design/proactive-diagnosis.md:96]: Can we distinguish the hang  checkpoint or dataloader  hang with NCCL OP hang?

File: docs/design/proactive-diagnosis.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+training operations (it can only cover some common cases).
+
+Therefore, a comprehensive, scalable framework is needed to gradually improve 
+the diagnosis and handling of corresponding exception scenarios.
+
+## Target
+
+- Be capable of proactive probing.
+- Requires flexible and scalable implementation.
+- Be able to reuse the current fault tolerance mechanisms.
```
Comment (workingloong) [inline docs/design/proactive-diagnosis.md:27]: Can you list which failure we can diagnose?

File: docs/design/proactive-diagnosis.md (hunk 1 above)
Comment (BalaBalaYi) [reply docs/design/proactive-diagnosis.md:68]: Not for now. Reason:

1. Need version >= 2.5.
2. Still a prototype feature.
3. The feature needs extra work if using Ascend device.

File: docs/design/proactive-diagnosis.md (hunk 2 above)
Comment (BalaBalaYi) [reply docs/design/proactive-diagnosis.md:96]: The XPU-Timer is still gradually being optimized in determining whether it is waiting for a barrier or experiencing an NCCL hang.

File: docs/design/proactive-diagnosis.md (hunk 3 above)
Comment (BalaBalaYi) [reply docs/design/proactive-diagnosis.md:27]: The currently implemented active diagnosis:
1.  heartbeat timeout 
2. training hang. 

Plan to refactor the previously implemented resource monitor into this framework, adding new diagnostic logic based on resource metrics.

Comment (BalaBalaYi) [reply docs/design/proactive-diagnosis.md:27]: Doc updated.

File: docs/design/proactive-diagnosis.md (hunk 1 above)
Comment (workingloong) [reply docs/design/proactive-diagnosis.md:68]: Torch 2.4.0 works.

**Pull request merged**